src/XliffTasks/Model/XlfDocument.cs

```diff
@@ -168,19 +169,75 @@ public bool Update(TranslatableDocument sourceDocument, string sourceDocumentId)
                     continue;
                 }
 
-                bodyElement.Add(
+                XElement newTransUnit = 
                     new XElement(ns + "trans-unit",
                         new XAttribute("id", sourceNode.Id),
                         new XElement(ns + "source", sourceNode.Source),
                         new XElement(ns + "target", new XAttribute("state", "new"), sourceNode.Source),
-                        new XElement(ns + "note", sourceNode.Note == "" ? null : sourceNode.Note)));
+                        new XElement(ns + "note", sourceNode.Note == "" ? null : sourceNode.Note));
+
+                bool inserted = false;
+                foreach (var transUnit in bodyElement.Elements(ns + "trans-unit"))
+                {
+                    if (StringComparer.Ordinal.Compare(newTransUnit.Attribute("id").Value, transUnit.Attribute("id").Value) < 0)
```

Comment:
This is going to be `O(N*M)` where N is the number of translation units in the already in the file and M is the number of source nodes. Perhaps nodesById should be a sorted list and we could find the node to insert before in log(N) time.

Reply:
Wait, nodes by id are the source nodes.

How about this: sort the left over new nodes and do the equivalent of merge sort merge.

Reply:
At this point we're only handling _new_ nodes, so in practice M should be very small.

Reply:
You're right that M is new nodes. First use of xlf in an existing project is one real case where it can be large. I'm just generally allergic to nested loops over arbitrary inputs like this.